File: visual-builder/src/types/execution.ts
/**
 * Execution type definitions for Visual Builder
 *
 * This module defines execution states, modes, and result structures
 * for workflow runtime operations.
 */

/**
 * Execution status lifecycle states
 */
export type ExecutionStatus =
  | 'pending'
  | 'queued'
  | 'running'
  | 'paused'
  | 'completed'
  | 'failed'
  | 'cancelled'
  | 'retrying'
  | 'timed_out';

/**
 * Execution mode options
 */
export type ExecutionMode = 'full' | 'mock' | 'debug';

/**
 * Workflow trigger types
 */
export type TriggerType = 'manual' | 'cron' | 'webhook';

/**
 * Individual node execution result
 *
 * Tracks the execution state and result of a single node
 * within a workflow execution.
 */
export interface NodeExecution {
  /** ID of the node that was executed */
  nodeId: string;
  /** Current status of this node's execution */
  status: ExecutionStatus;
  /** Input data passed to the node */
  input: unknown;
  /** Output data produced by the node, null if not completed */
  output: unknown | null;
  /** Error message if execution failed, null otherwise */
  error: string | null;
  /** ISO timestamp when node execution started */
  startedAt: string;
  /** ISO timestamp when node execution completed, null if still running */
  completedAt: string | null;
  /** Duration of execution in milliseconds, null if not completed */
  durationMs: number | null;
  /** Number of retry attempts for this node */
  retryCount: number;
}

/**
 * Complete workflow execution record
 *
 * Contains all information about a workflow execution run including
 * metadata, results, and individual node execution states.
 */
export interface Execution {
  /** Unique execution ID */
  id: string;
  /** ID of the workflow that was executed */
  workflowId: string;
  /** Current overall execution status */
  status: ExecutionStatus;
  /** Execution mode used */
  mode: ExecutionMode;
  /** Type of trigger that initiated this execution */
  triggerType: TriggerType;
  /** ID of the trigger if applicable, null for manual triggers */
  triggerId: string | null;
  /** Input data provided to the workflow */
  input: string;
  /** Final output data from the workflow, null if not completed */
  output: unknown | null;
  /** Error message if execution failed, null otherwise */
  error: string | null;
  /** Array of individual node executions */
  nodeExecutions: NodeExecution[];
  /** ISO timestamp when workflow execution started */
  startedAt: string;
  /** ISO timestamp when workflow execution completed, null if still running */
  completedAt: string | null;
  /** Total duration of execution in milliseconds, null if not completed */
  durationMs: number | null;
  /** Total LLM tokens used */
  totalTokens: number | null;
  /** Prompt/input tokens used */
  promptTokens: number | null;
  /** Completion/output tokens used */
  completionTokens: number | null;
  /** Estimated cost in USD */
  estimatedCost: number | null;
  /** LLM model used for execution */
  modelUsed: string | null;
}

File: visual-builder/backend/app/models/execution.py
"""Execution model."""
from datetime import UTC, datetime
from sqlalchemy import String, Text, Integer, DateTime, ForeignKey, Index
from sqlalchemy.orm import Mapped, mapped_column
from app.db.database import Base


def _utcnow():
    """Return current UTC time without timezone info (for SQLite compat)."""
    return datetime.now(UTC).replace(tzinfo=None)


class Execution(Base):
    """Execution entity."""

    __tablename__ = "executions"
    __table_args__ = (
        Index("ix_executions_workflow_status", "workflow_id", "status"),
        Index("ix_executions_workflow_started", "workflow_id", "started_at"),
    )

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
    )
    workflow_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("workflows.id"),
        nullable=False,
    )
    status: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        default="pending",
        index=True,
    )
    mode: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        default="full",
    )
    trigger_type: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        default="manual",
    )
    trigger_id: Mapped[str | None] = mapped_column(
        String(36),
        nullable=True,
    )
    input: Mapped[str] = mapped_column(
        Text,
        nullable=False,
        default="{}",
    )
    output: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )
    error: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )
    node_logs: Mapped[str] = mapped_column(
        Text,
        nullable=False,
        default="[]",
    )
    started_at: Mapped[datetime] = mapped_column(
        DateTime,
        nullable=False,
        default=_utcnow,
        index=True,
    )
    completed_at: Mapped[datetime | None] = mapped_column(
        DateTime,
        nullable=True,
    )
    duration_ms: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )
    total_tokens: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )
    prompt_tokens: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )
    completion_tokens: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )
    estimated_cost: Mapped[float | None] = mapped_column(
        nullable=True,
    )
    model_used: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
    )
